### PR TITLE
Add error handling and logging to git service

### PR DESCRIPTION
**Pull request by nully0x**
This PR focused on adding possible errors for the git-service where possible.  This partly closes issue #2

### CODE REVIEW

**Comment by nully0x** [comment]
This completes implementation for issue #2

File: git-service/src/softserve/handlers.rs [hunk 1]
```diff
@@ -1,32 +1,103 @@
...
+    execute_command(command)
+        .map(|output| {
+            let users: Vec<&str> = output.lines().collect();
+            Ok(users.iter().any(|&user| user.trim() == username))
+        })
+        .map_err(|e| {
+            error!("Failed to check user existence: {}", e);
+            GitServiceError::FailedToCheckUserExistence(e.to_string())
+        })?
+}
```
Comment (daviroo) [inline git-service/src/softserve/handlers.rs]: The `?` operator at the end of these functions is unnecessary and incorrect. 

The functions are already returning a `Result`, so using `?` here doesn't make sense. It's trying to unwrap a `Result` that's meant to be the return value. 

Remove the `?` and the `Ok()` wrapper in the `map()` closure to fix this. The `map_err()` chain will ensure the correct `Result` type is returned.

This change will preserve the error-handling logic while correctly adhering to the function's return type.
  ```rust
  pub fn check_repo_exists(repo_name: &str) -> Result<bool, GitServiceError> {
    let command = "repo list";
    execute_command(command)
        .map(|output| {
            let repos: Vec<&str> = output.lines().collect();
            repos.iter().any(|&repo| repo.split_whitespace().any(|word| word == repo_name))
        })
        .map_err(|e| {
            error!("Failed to check repo existence: {}", e);
            GitServiceError::FailedToCheckRepoExistence(e.to_string())
        })
}

pub fn check_user_exists(username: &str) -> Result<bool, GitServiceError> {
    let command = "user list";
    execute_command(command)
        .map(|output| {
            let users: Vec<&str> = output.lines().collect();
            users.iter().any(|&user| user.split_whitespace().any(|word| word == username))
        })
        .map_err(|e| {
            error!("Failed to check user existence: {}", e);
            GitServiceError::FailedToCheckUserExistence(e.to_string())
        })
}
  ```

File: git-service/src/softserve/mod.rs [hunk 2]
```diff
@@ -1,17 +1,46 @@
...
+    repo_url: &str,
+) -> Result<CreateRepoResponse, Box<dyn std::error::Error>> {
+    let import_command = format!("repo import {} {}", repo_name, repo_url);
+    execute_command(&import_command)?;
+
+    let token_command = format!("token create {}_token", repo_name);
+    let token_output = execute_command(&token_command)?;
+    let token = token_output.trim().to_string();
+
+    let server_url = config::get_server_url();
```
Comment (daviroo) [inline git-service/src/softserve/mod.rs]: There's a minor issue with how the config module is being referenced in the create_repo function. To ensure we're using the correct config module (the one declared as a submodule in this file), we should use the self keyword to reference it.

```suggestion
    let server_url = self::config::get_server_url();
```

File: README.md [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+```bash
+git clone [https://github.com/Extheoisah/hxckr-infra.git](https://github.com/Extheoisah/hxckr-infra.git)
+cd hxckr-infra
+```
+
+Run `cp .env.example .env` to create a `.env` file.
+> Adjust these values as needed to match your configuration.
+
+### 3. Configure Nix
```
Comment (daviroo) [inline README.md]: Very minor but we are missing a step 2 here.

**Review by daviroo — CHANGES_REQUESTED** [review]
Just some minor suggestions.

**Review by daviroo — APPROVED** [review]

File: README.md
Comment (Extheoisah) [inline README.md:1]: This readme is specific to setting up the git service. It should live inside the git service folder. The readme in the root should be one that explains the setup of the whole infra not just a subset

File: README.md [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+> We use flakes to manage the project dependencies. To enable flakes after installing nix, add
+```bash
+experimental-features = nix-command flakes
+```
+> to your `~/.config/nix/nix.conf` file.
+
+
+### 3. Set up Soft Serve
+Install Soft Serve:
+> for development, soft serve executable have been provide in the nix setup for the project you don't need to install it separately.
```
Comment (Extheoisah) [inline README.md]:
```suggestion
For development, a soft serve executable has been provided in the nix setup for the project so that you don't have to install it separately.
```

File: README.md [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+### 3.1. Before initializing Soft Serve
+> Before initializing Soft Serve, you need to set up the SSH key that will be used to authenticate with the server. This key should be added to the authorized keys in Soft Serve.
```
Comment (Extheoisah) [inline README.md]: nit: no need to wrap this in a comment block
```suggestion
Before initializing Soft Serve, you need to set up the SSH key that will be used to authenticate with the server. This key should be added to the authorized keys in Soft Serve.
```

File: README.md [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+> Generate new ed_25519 key pair using the following command:
+
+```bash
+ ssh-keygen -t ed25519 -C "your email or any identifier"
+```
+> This will generate a new SSH key pair in the default location (`~/.ssh/id_ed25519` and `~/.ssh/id_ed25519.pub`).
```
Comment (Extheoisah) [inline README.md]: nit: no need to wrap this in a comment block

```suggestion
This will generate a new SSH key pair in the default location (`~/.ssh/id_ed25519` and `~/.ssh/id_ed25519.pub`).
```

File: README.md [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+> View the public key using the following command:
```
Comment (Extheoisah) [inline README.md]: nit: no need to wrap this in a comment block

```suggestion
View the public key using the following command:
```

File: README.md [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+```bash
+cat ~/.ssh/id_ed25519.pub
+```
+> Copy the public key and add it to `SOFT_SERVE_INITIAL_ADMIN_KEYS` in your .env file.
```
Comment (Extheoisah) [inline README.md]: nit: no need to wrap this in a comment block
```suggestion
Copy the public key and add it to `SOFT_SERVE_INITIAL_ADMIN_KEYS` in your .env file.
```

File: README.md [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+> Next is to set config to use your private key for the server. Add the following to your `~/.ssh/config` file:
```
Comment (Extheoisah) [inline README.md]: nit: no need to wrap this in a comment block
```suggestion
Next is to set config to use your private key for the server. Add the following to your `~/.ssh/config` file:
```

File: README.md [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+```bash
+Host soft
+  HostName localhost
+  Port 23231
+  IdentityFile ~/.ssh/id_ed25519
+  IdentitiesOnly yes
+```
+Initialize Soft Serve:
```
Comment (Extheoisah) [inline README.md]: nit: add newline
```suggestion

Initialize Soft Serve:
```

File: README.md [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+```bash
+soft serve
+```
+> This will start the Soft Serve server on `localhost:23231`.
```
Comment (Extheoisah) [inline README.md]: nit: no need to wrap this in a comment block and add newline after sentence
```suggestion
This will start the Soft Serve server on `localhost:23231`.

```

File: README.md [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+### 3.2. After initializing Soft Serve
+> After initializing Soft Serve, the default DB is SQLite.
```
Comment (Extheoisah) [inline README.md]: We can reword this to:
```suggestion
The default DB for soft serve is SQLite.
```

File: README.md [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+> You can change the DB to Postgres by setting the `SOFT_SERVE_POSTGRES_DB` in the `.env` file.
```
Comment (Extheoisah) [inline README.md]:
```suggestion
However, you can change the DB to Postgres by setting the `SOFT_SERVE_POSTGRES_DB` in the `.env` file. We use Postgres as our db for this project.
```

File: README.md [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+  # Valid values are "sqlite" and "postgres".
+  driver: "postgres"
+  # The database data source name.
+  # This is driver specific and can be a file path or connection string.
+  # Make sure foreign key support is enabled when using SQLite.
+  data_source: "postgres://soft_serve:soft_serve@0.0.0.0:5432/soft_serve?sslmode=disable"
+```
+
+
+## Setup git-service
```
Comment (Extheoisah) [inline README.md]:
```suggestion
## Setup Git Service
```

File: README.md [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+### Build and Run the Project
+In base directory of the project, run the following commands(also exports all the env variable needed for development):
```
Comment (Extheoisah) [inline README.md]: nit
```suggestion
In base directory of the project, run the following commands (also ensure to have all the environment variables needed for development in the `.env` file. You can find a sample in the `.env.sample` file):
```

File: README.md [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+```bash
+nix develop
+```
+
+Start the soft serve server in nix shell:
+```bash
+soft serve
+```
+
+In another nix shell(you enter nix shell after running `nix develop` from the codebase base directory) cd into the `git-service` directory:
```
Comment (Extheoisah) [inline README.md]:
```suggestion
In another nix shell (you enter nix shell after running `nix develop` from the codebase base directory) cd into the `git-service` directory:
```

File: git-service/src/softserve/handlers.rs [hunk 17]
```diff
@@ -1,32 +1,103 @@
...
+use super::{create_repo, create_token, create_user};
+use actix_web::{web, HttpResponse, Result};
+use serde_json::json;
+use tracing::{error, info, warn};
+
+pub async fn handle_create_user(
+    user: web::Json<CreateUserRequest>,
+) -> Result<HttpResponse, GitServiceError> {
+    if check_user_exists(&user.username)? {
+        warn!("Attempt to create existing user: {}", user.username);
```
Comment (Extheoisah) [inline git-service/src/softserve/handlers.rs]: This log should be an error and not a warning since it returns an error
```suggestion
        error!("Attempt to create an existing user: {}", user.username);
```

File: git-service/src/softserve/handlers.rs [hunk 18]
```diff
@@ -1,32 +1,103 @@
...
-pub async fn test_connection() -> impl Responder {
-    match create_user("test_user") {
-        Ok(_) => HttpResponse::Ok().body("Connection to Soft Serve successful"),
-        Err(e) => HttpResponse::InternalServerError().body(format!("Connection failed: {}", e)),
+pub fn check_repo_exists(repo_name: &str) -> Result<bool, GitServiceError> {
+    let command = "repo list";
+    execute_command(command)
+        .map(|output| {
+            let repos: Vec<&str> = output.lines().collect();
+            repos.iter().any(|&repo| repo.trim() == repo_name)
```
Comment (Extheoisah) [inline git-service/src/softserve/handlers.rs]: You may want to convert the repo_name to lowercase before validation. We should do the same when saving the repo name in the database
```suggestion
            repos.iter().any(|&repo| repo.to_lowercase().trim() == repo_name.trim().to_lowercase())
```

File: git-service/src/softserve/handlers.rs [hunk 19]
```diff
@@ -1,32 +1,103 @@
...
+            GitServiceError::FailedToCheckRepoExistence(e.to_string())
+        })
+}
+
+pub fn check_user_exists(username: &str) -> Result<bool, GitServiceError> {
+    let command = "user list";
+    execute_command(command)
+        .map(|output| {
+            let users: Vec<&str> = output.lines().collect();
+            users.iter().any(|&user| user.trim() == username)
```
Comment (Extheoisah) [inline git-service/src/softserve/handlers.rs]: You may want to convert the username to lowercase before validation. We should do the same when saving the username in the database
```suggestion
            users.iter().any(|&user_name| user_name.to_lowercase().trim() == username.trim().to_lowercase())
```

File: git-service/src/softserve/handlers.rs [hunk 20]
```diff
@@ -1,32 +1,103 @@
...
+            error!("Failed to check user existence: {}", e);
+            GitServiceError::FailedToCheckUserExistence(e.to_string())
+        })
+}
+
+pub async fn handle_create_repo(
+    repo_req: web::Json<CreateRepoRequest>,
+) -> Result<HttpResponse, GitServiceError> {
+    if check_repo_exists(&repo_req.repo_name)? {
+        warn!("Attempt to create existing repo: {}", repo_req.repo_name);
```
Comment (Extheoisah) [inline git-service/src/softserve/handlers.rs]: This log should be an error and not a warning since it returns an error
```suggestion
        error!("Attempt to create an existing repo: {}", repo_req.repo_name);
```

**Review by Extheoisah — COMMENTED** [review]
PR looks good. My reviews are mostly nits

**Review by Extheoisah — APPROVED** [review]
